src/hotspot/os/linux/cgroupUtil_linux.cpp

```diff
@@ -46,3 +47,100 @@ int CgroupUtil::processor_count(CgroupCpuController* cpu_ctrl, int host_cpus) {
+    // update to shortened path and try again
+    mem->set_subsystem_path(cg_path);
+    limit = mem->read_memory_limit_in_bytes(phys_mem);
+    path_iterated = true;
+    if (limit > 0) {
+      log_trace(os, container)("Adjusted controller path for memory to: %s", mem->subsystem_path());
+      os::free(cg_path);
+      os::free(orig);
+      return;
+    }
+  }
+  // no lower limit found or limit at leaf
+  os::free(cg_path);
+  if (path_iterated) {
+    mem->set_subsystem_path("/");
+    limit = mem->read_memory_limit_in_bytes(phys_mem);
+    if (limit > 0) {
+      // handle limit set at mount point
+      log_trace(os, container)("Adjusted controller path for memory to: %s", mem->subsystem_path());
+      os::free(orig);
+      return;
+    }
+    log_trace(os, container)("No lower limit found in hierarchy %s, adjusting to original path %s",
+                              mem->mount_point(), orig);
+    mem->set_subsystem_path(orig);
+  } else {
+    log_trace(os, container)("Lowest limit for memory at leaf: %s",
+                              mem->subsystem_path());
+  }
+  os::free(orig);
+}
+
+void CgroupUtil::adjust_controller(CgroupCpuController* cpu) {
+  if (!cpu->needs_hierarchy_adjustment()) {
+    // nothing to do
+    return;
+  }
+  log_trace(os, container)("Adjusting controller path for cpu: %s", cpu->subsystem_path());
+  assert(cpu->cgroup_path() != nullptr, "invariant");
+  char* orig = os::strdup(cpu->cgroup_path());
+  char* cg_path = os::strdup(orig);
+  char* last_slash;
+  assert(cg_path[0] == '/', "cgroup path must start with '/'");
+  int host_cpus = os::Linux::active_processor_count();
+  int cpus = CgroupUtil::processor_count(cpu, host_cpus);
+  bool path_iterated = false;
+  while (cpus == host_cpus && (last_slash = strrchr(cg_path, '/')) != cg_path) {
+    *last_slash = '\0'; // strip path
+    // update to shortened path and try again
+    cpu->set_subsystem_path(cg_path);
+    cpus = CgroupUtil::processor_count(cpu, host_cpus);
+    path_iterated = true;
+    if (cpus != host_cpus) {
+      log_trace(os, container)("Adjusted controller path for cpu to: %s", cpu->subsystem_path());
+      os::free(cg_path);
+      os::free(orig);
+      return;
+    }
+  }
+  // no lower limit found or limit at leaf
+  os::free(cg_path);
+  if (path_iterated) {
+    cpu->set_subsystem_path("/");
+    cpus = CgroupUtil::processor_count(cpu, host_cpus);
+    if (cpus != host_cpus) {
+      // handle limit set at mount point
+      log_trace(os, container)("Adjusted controller path for cpu to: %s", cpu->subsystem_path());
+      os::free(orig);
+      return;
+    }
+    log_trace(os, container)("No lower limit found in hierarchy %s, adjusting to original path %s",
+                              cpu->mount_point(), orig);
+    cpu->set_subsystem_path(orig);
+  } else {
+    log_trace(os, container)("Lowest limit for cpu at leaf: %s",
+                              cpu->subsystem_path());
+  }
+  os::free(orig);
+  return;
+}
```

Comment:
The duplications are regrettable. We usually do these things with a Closure object, but that is hard to do here, among other things since the controller object itself is mutable.

Reply:
FYI @tstuefe  [I did unify](https://github.com/openjdk/jdk/compare/82db527b74a3dc49ea7dc05632be806a6c193770...fc17369cfb080b525d45deec3a2956d704851a69.diff) that @jerboaa's code by a template but @jerboaa did not like it. ([My patch](https://openjdk.github.io/cr/?repo=jdk&pr=17198&range=11) had neither code duplications nor templates.)